### DISCUSSION TITLE
Proper usage of "shadingMode" arg (during import)

### DISCUSSION THREAD

**Discussion by pmolodo** [Autodesk/maya-usd#1131] [Q&A]
Hi - I was a bit confused by the new vector-of-vectors `shadingMode` arg.  Is there somewhere that explains what each of the args represents?  And how it relates to the values output by `cmds.mayaUSDListShadingModes(im=1)`?

**Comment by JGamache-autodesk** [comment]
When assembling scenes with multiple assets, they can have been exported with multiple shadingModes. When importing the assembly we previously searched only for a single shading mode and might miss a few materials.

The vector allows searching more intensively for shadingModes and build appropriate Maya materials for each one that is found. The shading modes are searched in the order specified by the vector, and the first shadingMode that results in a Material ends the search.

Getting back the previous behavior involves specifying only one shadingMode in the vector.

**Reply by pmolodo** [reply]
Gotcha, that makes sense - what do the pair of elements in the 2nd dimension represent?

**Reply by JGamache-autodesk** [reply]
We have shadingModes, which we would like to phase out in favor of the more flexible `useRegistry` shading mode. So we currently need a 2nd dimension to specify the `convertMaterialFrom` part of a `useRegistry` shading mode (we have pxrRis and  UsdPreviewSurface in the code, with MaterialX and others as future candidates).

If we manage to phase out all of the old shadingModes modes (like the old pxrRis) without introducing new ones, we will be able to remove that second dimension since the first element will always be `useRegistry`.

**Reply by JGamache-autodesk** [reply]
The documentation in `plugin\pxr\doc\README.md` is up to date, what is missing is creating its equivalent on the mayaUsdPlugin documentation side.

**Reply by pmolodo** [reply]
Ah! Thanks for the doc pointer - that does a good job of describing the vector, and the first value of the pair.

Unfortunately, that doc is still a little unclear on the meaning of the 2nd term... from a combination of your responses, that doc, and playing with `mayaUSDListShadingModes`, I'm gathering that:

- The second term is only ever really needed if the first term is `'useRegistry'`
- If the first term is anything other than `'useRegistry'`, the second value should always be `'none'`...?
- Conceptually, a "shadingMode" is associated with a fixed PAIR of values
   - ie, `"My Shading Mode" == ("useRegistry", "myShadingMode")`

Is this correct?

**Reply by JGamache-autodesk** [reply]
Yes, this is correct. 
As you saw with "mayaUSDListShadingModes", we try the most advanced modes first, which means we start with any "useRegistry" mode that is not preview surface then we try preview surface, then finally we go to the legacy modes like displayColor when all else fails.

**Reply by pmolodo** [reply]
Ok.  Well, now that I have a clearer idea of how the current system works - I'd like to suggest that we could simplify it.  While the two-value option might mirror how the code flow works behind the scenes, it's not very intuitive, or how an average user / artist would think about it.  And I think we can simplify, because one of the two options is always "fixed".

That is, a given shading mode is always either a `useRegistry` style shading more or not:

- if it is `useRegistry`, then the first value is always `"useRegistry"`
- if it isn't `useRegistry`, then the second value is always `"none"`

So, our current two-arg options:

```
[''useRegistry', 'renderManForMaya']
['pxrRis', 'none']
['useRegistry', 'UsdPreviewSurface']
['displayColor', 'none']
['none', 'none']
```

...could just become the much-simpler:

```
'renderManForMaya'
'pxrRis'
'UsdPreviewSurface'
'displayColor'
'none'
```

We do need to know whether a given shadingMode is `useRegistry` or not, but that could simply become a boolean option, defined on each individual shadingMode. It would not need to be explicitly exposed to end users.

Does this make sense?  Is there a reason why this wouldn't work that I'm missing?

**Reply by JGamache-autodesk** [reply]
Makes perfect sense. As long as there is no overlap between the two sets there is no reason to keep this second dimension.
